File: backend/services/agent-runner/worker/activities/graphton/prompt_builder.py
"""System-prompt construction for Graphton agent execution.

All functions are pure: no side effects, no state, no I/O beyond
``os.path`` calls for workspace-relative file metadata.

Extracted from ``execute_graphton.py`` to keep prompt logic focused
and independently testable.
"""

from __future__ import annotations

import os
from typing import Any

from worker.workspace import GitMetadata, ProvisionResult, SourceType
from worker.workspace.tree import (
    TREE_DEFAULT_MAX_ENTRIES as _TREE_MAX_ENTRIES,
)
from worker.workspace.tree import (
    build_directory_tree as _build_directory_tree,
)
from worker.workspace.tree import (
    human_size as _human_size,
)


def build_workspace_prompt_section(
    provision_results: list[ProvisionResult],
    *,
    container_root: str = "",
) -> str:
    """Build the ``## Workspace`` system prompt section from provision results.

    Returns an empty string when there are no provisions — safe to
    unconditionally append.

    For a single entry the output uses the legacy single-workspace format.
    For multiple entries each gets a ``### {name}`` sub-heading.

    *container_root* is the backend's root directory — used only in the
    multi-entry path to tell the agent its current working directory.
    """
    if not provision_results:
        return ""

    if len(provision_results) == 1:
        return _build_single_workspace_section(provision_results[0])

    return _build_multi_workspace_section(provision_results, container_root)


def _git_writeback_guidance(
    meta: GitMetadata | None,
    *,
    heading_level: int = 3,
) -> str:
    """Return a prompt section telling the agent it can push changes.

    Returns an empty string when *meta* is ``None`` or credentials
    were not configured.
    """
    if meta is None or not meta.git_credentials_configured:
        return ""

    heading = "#" * heading_level
    return (
        f"\n\n{heading} Git Write-Back\n\n"
        "Git credentials are configured — you can push changes to "
        "the remote repository.\n\n"
        "**Rules:**\n"
        "- Create a new branch for your changes (never push directly "
        "to the default branch).\n"
        "- Write clear, meaningful commit messages.\n"
        "- Push your branch and report the branch name when done.\n"
        "- After pushing, use `create_pull_request` to open a PR. "
        "It reads credentials and repo info automatically.\n"
        "- Do NOT read, echo, or reference credential files "
        "(e.g. `~/.git-credentials`)."
    )


def _build_single_workspace_section(result: ProvisionResult) -> str:
    """Format the workspace section for a single entry (legacy compat)."""
    if not result.workspace_description:
        return ""

    section = "\n\n## Workspace\n\n" + result.workspace_description

    if result.file_tree:
        section += "\n\n" + result.file_tree

    section += _git_writeback_guidance(result.git_metadata)

    return section


def _build_multi_workspace_section(
    results: list[ProvisionResult],
    container_root: str,
) -> str:
    """Format the workspace section for multiple entries.

    The tree heading level is controlled at provisioning time via
    ``tree_heading_level``, so no post-hoc replacement is needed.
    """
    first_label = results[0].entry_name or "entry-1"

    section = (
        f"\n\n## Workspace\n\n"
        f"This session has {len(results)} workspace entries.\n\n"
        f"**Current working directory**: `{container_root}`\n"
        f"**Path resolution**: All file tools (read, write, edit, ls, "
        f"glob, grep) resolve paths relative to the current working "
        f"directory. Use entry-relative paths "
        f"(e.g., `{first_label}/src/main.py`) or absolute paths.\n"
    )

    for idx, result in enumerate(results):
        label = result.entry_name or f"entry-{idx + 1}"
        section += f"\n### {label} (`{result.root_dir}`)\n\n"
        section += _format_entry_description(result)

        if result.file_tree:
            section += "\n\n" + result.file_tree

    return section


def _format_entry_description(result: ProvisionResult) -> str:
    """Generate a multi-workspace-appropriate description for one entry.

    Uses structured fields rather than the generic ``workspace_description``
    so the phrasing fits a multi-entry context.
    """
    name = result.entry_name or "this entry"

    if result.source_type == SourceType.LOCAL_PATH:
        return (
            f"Workspace entry **{name}** is the user's project directory "
            f"at `{result.root_dir}`.\n"
            "You are operating directly on the user's files — changes are "
            "immediate and persistent. Use git to track and verify your "
            "changes."
        )

    if result.source_type == SourceType.GIT_REPO and result.git_metadata:
        meta = result.git_metadata
        short_sha = (
            meta.base_commit[:7]
            if len(meta.base_commit) >= 7
            else meta.base_commit
        )
        desc = (
            f"Workspace entry **{name}** was initialized from "
            f"{meta.repo_url} (branch: {meta.branch}, "
            f"commit: {short_sha}).\n"
            "Changes you make will be captured as artifacts when "
            "execution completes."
        )
        desc += _git_writeback_guidance(meta, heading_level=4)
        return desc

    if result.source_type == SourceType.EMPTY:
        return (
            f"Workspace entry **{name}** is an empty workspace.\n"
            "Create files and directories as needed for your task."
        )

    return result.workspace_description


def build_referenced_files_prompt_section(
    workspace_file_refs: list[str],
    workspace_root: str,
) -> str:
    """Build the ``## Referenced Files`` system prompt section.

    When the user attaches workspace files, this lists them with structural
    metadata so the agent can navigate efficiently.
    """
    if not workspace_file_refs:
        return ""

    section = (
        "\n\n## Referenced Files\n\n"
        "The user has highlighted the following workspace paths for your "
        "attention. Use `read` to access file contents.\n\n"
    )

    for ref_path in workspace_file_refs:
        full_path = os.path.join(workspace_root, ref_path)
        try:
            if os.path.isdir(full_path):
                tree_lines, total = _build_directory_tree(
                    full_path,
                    ref_path.rstrip("/") + "/",
                )
                label = "entry" if total == 1 else "entries"
                section += f"- `{ref_path}/` (directory, {total} {label})\n"
                for line in tree_lines:
                    section += line + "\n"
                if total > len(tree_lines):
                    section += (
                        f"    - ... and {total - len(tree_lines)} more "
                        f"(truncated at {_TREE_MAX_ENTRIES} entries)\n"
                    )
            else:
                size = os.path.getsize(full_path)
                section += f"- `{ref_path}` ({_human_size(size)})\n"
        except OSError:
            section += f"- `{ref_path}`\n"

    return section


_RESPONSE_RULES = (
    "\n\n## Response rules\n\n"
    "- After using the read tool, NEVER reprint, echo, list, or summarize "
    "file contents in your response. Tool results are already in your "
    "context. Proceed directly to analysis or the task.\n"
    "- Do not begin responses with phrases like "
    '"Below is the complete content", '
    '"Here are the contents of the files", or similar. '
    "The user did not ask you to display file contents.\n"
)

_SUB_AGENT_RULES = (
    "\n\n## Sub-agent delegation rules\n\n"
    "- **Read files directly.** When you need the contents of a file, "
    "use the `read` tool yourself. Do not delegate file reading to "
    "sub-agents via the `task` tool. You need raw file contents in "
    "your own context to reason about them accurately.\n"
    "- Sub-agents are for **multi-step, independent tasks** that "
    "produce a deliverable (analysis, synthesis, generated content). "
    "They are not for fetching data that you will process yourself.\n"
    "- When delegating to a sub-agent, specify the analysis or "
    'deliverable you need — not "read these files and give me the '
    'contents."\n'
)


def enhance_system_prompt(
    *,
    instructions: str,
    provision_results: list[ProvisionResult],
    container_root: str,
    user_message: str,
    build_relevance: Any,
    workspace_roots: list[Any],
    skills_prompt_section: str,
    workspace_file_refs: list[str],
    workspace_root: str,
    injected_files: list[dict[str, Any]],
) -> str:
    """Assemble the full system prompt from base instructions and contextual sections.

    Pure function: collects all prompt parts and returns the final string.
    """
    prompt = instructions

    workspace_section = build_workspace_prompt_section(
        provision_results, container_root=container_root,
    )
    if workspace_section:
        prompt += workspace_section

    relevance_section = build_relevance(user_message, workspace_roots)
    if relevance_section:
        prompt += relevance_section

    if skills_prompt_section:
        prompt += skills_prompt_section

    if workspace_file_refs:
        ref_section = build_referenced_files_prompt_section(
            workspace_file_refs, workspace_root,
        )
        if ref_section:
            prompt += ref_section

    if injected_files:
        input_files_section = "\n\n## Input Files\n\n"
        input_files_section += (
            "The following files have been provided as read-only reference "
            "material for your task. They live under `.stigmer/inputs/` and "
            "are NOT part of the project source tree.\n\n"
            "Read them using the `read` tool when you need their contents. "
            "Do NOT echo, reprint, or summarize file contents in your response "
            "-- they are reference material, not output. "
            "Do NOT modify or delete these files.\n\n"
        )
        for f in injected_files:
            size_info = f" ({f['size']} bytes)" if f.get('size') is not None else ""
            input_files_section += f"- `{f['path']}`{size_info}\n"
        prompt += input_files_section

    prompt += _RESPONSE_RULES
    prompt += _SUB_AGENT_RULES

    return prompt

File: backend/services/runner/src/activities/execute-cursor/prompt-builder.ts
/**
 * Builds the enhanced prompt for the Cursor agent.
 *
 * Replicates prompt_builder.py::enhance_system_prompt() but delivers
 * the content via the first user message (Cursor SDK has no systemPrompt
 * parameter). On reinvocation, only the approval decisions are sent.
 *
 * Sections (in order, conditionally included):
 * 1. Interaction-mode / implement-plan directives
 * 2. Agent instructions (persona/character)
 * 3. Available skills metadata
 * 4. Sub-agent delegation guidance
 * 5. Workspace context (multi-root only; single-dir is redundant with SDK cwd)
 * 6. Input files / referenced files
 * 7. Response rules (only when agent has no custom instructions)
 * 8. User's actual message
 */

import { resolve } from "node:path";
import type { DatastoreUsage, SubAgent } from "@stigmer/protos/ai/stigmer/agentic/agent/v1/spec_pb";
import type { PendingApproval } from "@stigmer/protos/ai/stigmer/agentic/agentexecution/v1/approval_pb";
import { ApprovalAction, InteractionMode } from "@stigmer/protos/ai/stigmer/agentic/agentexecution/v1/enum_pb";
import { formatContextBridgeText } from "../../shared/context-bridge.js";
import { formatConversationCatchupText } from "../../shared/conversation-catchup.js";
import { formatDatastoresSection } from "../../shared/datastore-attachment.js";
import {
  formatChannelTemplatesSection,
  type ChannelMessagingInfo,
} from "../../shared/channel-attachment.js";
import {
  formatSenderIdentityText,
  type SenderIdentity,
} from "../../shared/sender-identity.js";
import { formatSessionContextText } from "../../shared/session-context.js";
import {
  visionDisclosureLines,
  type NotViewableEntry,
} from "../../shared/attachment-vision.js";
import { PLAN_MODE_DIRECTIVE } from "../../shared/plan-mode-prompt.js";
import {
  buildImplementPlanDirective,
  findApprovedPlanPath,
} from "../../shared/implement-plan-prompt.js";

/**
 * Marker path segments that identify runner-internal directories. A workspace
 * dir containing any of these segments is almost certainly the runner's own
 * app directory, not a user workspace.
 */
const RUNNER_PATH_MARKERS = [
  "/runtimes/cursor-runner/",
  "/runtimes/agent-runner/",
  "/node_modules/",
  "/dist/main.js",
] as const;

export interface SkillMetadata {
  name: string;
  description: string;
  path: string;
}

/**
 * Vision facts about this turn's attachments, rendered inside the
 * input-files section: which images the model can see inline (in send
 * order) and which degraded to the file-pointer story. Derived per turn
 * from the resolved attachments — never carried across turns.
 */
export interface VisionPromptInfo {
  inlineFilenames: string[];
  notViewable: NotViewableEntry[];
}

/**
 * One resolved attachment as the `<input_files>` section renders it —
 * the final workspace-relative path plus the duplicate-rename disclosure
 * (attachment-resolver.ts / shared/attachment-naming.ts).
 */
export interface AttachmentPromptEntry {
  /** Workspace-relative path the agent reads (`.stigmer/inputs/{filename}`). */
  path: string;
  /** Original filename, present only when a duplicate name was renamed. */
  renamedFrom?: string;
}

export interface EnhancedPromptOptions {
  instructions: string;
  userMessage: string;
  skills: SkillMetadata[];
  /**
   * Datastores attached via `datastore_usages` — rendered as the
   * `<available_datastores>` section (DD-005 SD-5, skills precedent)
   * pointing the model at the synthesized record tools. Always the healthy
   * rendering here: the Cursor SDK connects MCP itself, so this harness can
   * never observe the live roster — the section's standing
   * failure-disclosure instruction is its whole coverage (issue #325).
   */
  datastoreUsages?: DatastoreUsage[];
  /**
   * Serving proactive channels + their approved templates — rendered as
   * the `<available_channel_templates>` section (proactive-messaging
   * DD-003 D5) beside the synthesized send_channel_message tool.
   */
  channelMessaging?: ChannelMessagingInfo[];
  subAgents: SubAgent[];
  workspaceDirs: string[];
  workspaceFileRefs: string[];
  attachments: AttachmentPromptEntry[];
  /** Inline/degraded image facts for the input-files section (T04 vision). */
  vision?: VisionPromptInfo;
  interactionMode?: InteractionMode;
  /**
   * The execution is a Build-from-plan turn (spec.execution_config
   * .build_from_plan): inject the implement-plan directive so the model reads
   * the attached approved plan (or falls back to the conversation's plan).
   * The user message itself is just a short label ("Build from plan").
   */
  buildFromPlan?: boolean;
  /**
   * Rollover context bridge (cloud DD-013): a digest of the previous
   * session's conversation, read from `SessionSpec.metadata`. Lands in the
   * first message, so it persists in the cursor agent's own conversation
   * store for the session's lifetime — buildEnhancedPrompt firing only on
   * first/fresh executions is exactly the right delivery.
   */
  contextBridge?: string;
  /**
   * Channel sender identity (attribution, not authorization): the
   * provider-verified identifier of the person on the channel, read from
   * `SessionSpec.metadata`. Like the bridge, it lands in the first message
   * and persists in the cursor agent's own conversation store — identity
   * is constant per conversation (channel sessions are keyed per-sender).
   */
  senderIdentity?: SenderIdentity;
  /**
   * Embedder-supplied session context (personalization, not
   * authorization): standing free-text context about the user/session,
   * read from `SessionSpec.metadata`. Like the bridge, it lands in the
   * first message and persists in the cursor agent's own conversation
   * store — the context is constant for the session's lifetime.
   */
  sessionContext?: string;
  /**
   * Conversation catchup (cloud DD-006): what happened on the channel
   * conversation that the agent has not seen, read from the execution
   * spec's `conversation_catchup`. PER-TURN, unlike the three standing
   * siblings above: it rides BOTH prompt paths — this enhanced prompt and
   * a resumed turn's prefix (the `interaction_mode` shape) — because
   * handback lands mid-session on a resumed agent, the exact case the
   * metadata lane cannot reach.
   */
  conversationCatchup?: string;
}

/**
 * Build the enhanced prompt for the first execution in a session.
 *
 * Prepends agent blueprint information before the user's message so the
 * Cursor agent has full context. This avoids writing files to the workspace.
 */
export function buildEnhancedPrompt(options: EnhancedPromptOptions): string {
  const sections: string[] = [];

  const modePrefix = formatInteractionModePrefix(options.interactionMode);
  if (modePrefix) {
    sections.push(modePrefix);
  }

  const implementPlan = formatImplementPlanSection(
    options.buildFromPlan,
    options.attachments,
  );
  if (implementPlan) {
    sections.push(implementPlan);
  }

  if (options.instructions) {
    sections.push(formatInstructions(options.instructions));
  }

  if (options.skills.length > 0) {
    sections.push(formatSkillsSection(options.skills));
  }

  if (options.datastoreUsages !== undefined && options.datastoreUsages.length > 0) {
    sections.push(formatDatastoresSection(options.datastoreUsages));
  }

  if (options.channelMessaging !== undefined && options.channelMessaging.length > 0) {
    // "" when nothing is sendable — the tool alone still serves text
    // sends inside a 24-hour window (DD-006 D6).
    const channelSection = formatChannelTemplatesSection(options.channelMessaging);
    if (channelSection !== "") {
      sections.push(channelSection);
    }
  }

  if (options.subAgents.length > 0) {
    sections.push(formatSubAgentsSection(options.subAgents));
  }

  const safeDirs = sanitizeWorkspaceDirs(options.workspaceDirs);
  // Encourage Cursor's built-in `explore` sub-agent for codebase work whenever
  // there is a workspace to explore — the headless agent otherwise rarely
  // delegates exploration the way the Cursor IDE does.
  if (safeDirs.length > 0) {
    sections.push(formatExplorationGuidance());
  }
  if (safeDirs.length > 1) {
    sections.push(formatWorkspaceContext(safeDirs));
  }

  if (options.attachments.length > 0) {
    sections.push(formatInputFiles(options.attachments, options.vision));
  }

  if (options.workspaceFileRefs.length > 0) {
    sections.push(formatReferencedFiles(options.workspaceFileRefs));
  }

  if (!options.instructions) {
    sections.push(formatResponseRules());
  }

  // The sender identity precedes the bridge: it is standing context about
  // WHO the conversation is with, which the carried conversation may refer
  // back to.
  if (options.senderIdentity) {
    sections.push(formatSenderIdentitySection(options.senderIdentity));
  }

  // Standing facts about the user (session context) come before the
  // carried conversation (bridge): the bridge may refer back to them.
  if (options.sessionContext) {
    sections.push(formatSessionContextSection(options.sessionContext));
  }

  // The rollover context bridge sits directly before the protocol + task so
  // the carried conversation is the freshest CONTEXT the model reads —
  // while the approval protocol keeps its pinned last-before-task slot (it
  // is an INSTRUCTION and must outweigh everything, including this bridge).
  if (options.contextBridge) {
    sections.push(formatContextBridgeSection(options.contextBridge));
  }

  // Catchup after the bridge (DD-007 D-d: bridge first, catchup second) —
  // the bridge carries the pre-takeover conversation, the catchup the human
  // episode, strictly newer by construction; recency puts it closer to the
  // task.
  if (options.conversationCatchup) {
    sections.push(formatConversationCatchupSection(options.conversationCatchup));
  }

  // Always last before the task: the platform's tool-approval protocol. Placed
  // here for recency so it outweighs any "ask the user first" guidance Cursor
  // surfaces from a connected MCP server (see formatToolApprovalProtocol).
  sections.push(formatToolApprovalProtocol());

  sections.push(`<user_request>\n${options.userMessage}\n</user_request>`);

  return sections.join("\n\n---\n\n");
}

/**
 * Build the reinvocation prompt after approval decisions.
 *
 * Only contains approval instructions, no blueprint (it persists in the resumed
 * agent's native context). The actions are described in human-meaningful terms
 * (the resolved approval message, e.g. "Write file: foo.txt") — NOT the opaque
 * tool-call ids, which mean nothing to the model. A resumed agent re-issues the
 * approved tool with a fresh id; the approval gate (see approval-state.ts grants)
 * is what actually lets the re-attempt through.
 *
 * `appliedToolCallIds` are approved whole-file writes the RUNNER has already
 * applied to the workspace itself (exact-apply — the "what you approve is what
 * gets applied" guarantee). Those are described as ALREADY DONE so the model
 * does not redo them; only the remaining approved actions (hunk edits, shell,
 * MCP) are the ones it must still carry out. No grant is issued for an applied
 * write, so if the model edits that file again it is correctly re-gated.
 */
export function buildReinvocationPrompt(
  pendingApprovals: PendingApproval[],
  approvalDecisions: Map<string, ApprovalAction>,
  appliedToolCallIds: ReadonlySet<string> = new Set(),
): string {
  const approved: string[] = [];
  const alreadyApplied: string[] = [];
  const skipped: string[] = [];

  for (const pa of pendingApprovals) {
    const action = approvalDecisions.get(pa.toolCallId);
    const isApproved =
      action === ApprovalAction.APPROVE || action === ApprovalAction.APPROVE_ALL;
    if (isApproved && appliedToolCallIds.has(pa.toolCallId)) {
      alreadyApplied.push(describeApproval(pa));
    } else if (isApproved) {
      approved.push(describeApproval(pa));
    } else if (action === ApprovalAction.SKIP) {
      skipped.push(describeApproval(pa));
    }
  }

  const parts: string[] = [];
  if (alreadyApplied.length) {
    parts.push(
      "The user reviewed and APPROVED the following change(s), and the platform " +
        "has ALREADY applied them to the workspace exactly as shown. Do NOT redo " +
        "or rewrite them — treat them as done and continue with the rest of the " +
        "task:\n" +
        alreadyApplied.map((a) => `- ${a}`).join("\n"),
    );
  }
  if (approved.length) {
    parts.push(
      "The user reviewed the following action(s) you proposed and APPROVED them. " +
        "Carry them out now:\n" +
        approved.map((a) => `- ${a}`).join("\n"),
    );
  }
  if (skipped.length) {
    parts.push(
      "The user SKIPPED the following action(s). Do not perform them; continue " +
        "with the rest of the task without them:\n" +
        skipped.map((a) => `- ${a}`).join("\n"),
    );
  }

  // The model's "ask the user first" bias recurs every turn, including resumes,
  // so restate the protocol: keep invoking tools directly to finish the task —
  // the platform gates and resumes any further sensitive actions automatically.
  // Never ask for permission in prose. (See formatToolApprovalProtocol.)
  parts.push(
    "Continue the rest of the task by invoking the tools it requires directly. " +
      "The platform automatically requests approval for any further sensitive " +
      "action and resumes you — do not ask the user for permission in prose.",
  );

  return parts.join("\n\n");
}

/**
 * Render a pending approval as a short human-readable action description for the
 * reinvocation prompt. Prefers the already-resolved approval message; falls back
 * to the tool name plus a bounded args preview.
 */
function describeApproval(pa: PendingApproval): string {
  if (pa.message) return pa.message;
  const name = pa.mcpServerSlug ? `${pa.mcpServerSlug}/${pa.toolName}` : pa.toolName;
  if (!pa.argsPreview) return name;
  const preview = pa.argsPreview.length > 200
    ? `${pa.argsPreview.slice(0, 200)}…`
    : pa.argsPreview;
  return `${name} (${preview})`;
}

/**
 * Filters out workspace directories that appear to be runner-internal paths.
 * This prevents the prompt from exposing implementation details like
 * ~/.stigmer/runtimes/cursor-runner/... to the AI model.
 */
export function sanitizeWorkspaceDirs(dirs: string[]): string[] {
  const safe: string[] = [];

  for (const dir of dirs) {
    const resolved = resolve(dir);
    if (isRunnerInternalPath(resolved)) {
      console.warn(
        `Workspace dir "${dir}" appears to be a runner-internal path — ` +
        "omitting from prompt to prevent implementation detail leakage.",
      );
      continue;
    }
    safe.push(dir);
  }

  return safe;
}

function isRunnerInternalPath(absolutePath: string): boolean {
  return RUNNER_PATH_MARKERS.some((marker) => absolutePath.includes(marker));
}

export function formatInstructions(instructions: string): string {
  return `<agent_instructions>\n${instructions}\n</agent_instructions>`;
}

export function formatContextBridgeSection(bridge: string): string {
  return `<previous_conversation_context>\n${formatContextBridgeText(bridge)}\n</previous_conversation_context>`;
}

export function formatSenderIdentitySection(identity: SenderIdentity): string {
  return `<conversation_sender>\n${formatSenderIdentityText(identity)}\n</conversation_sender>`;
}

export function formatSessionContextSection(context: string): string {
  return `<session_context>\n${formatSessionContextText(context)}\n</session_context>`;
}

export function formatConversationCatchupSection(digest: string): string {
  return `<conversation_catchup>\n${formatConversationCatchupText(digest)}\n</conversation_catchup>`;
}

export function formatSkillsSection(skills: SkillMetadata[]): string {
  const entries = skills.map(
    (s) => `- **${s.name}**: ${s.description}\n  Path: \`${s.path}\``,
  );
  return [
    "<available_skills>",
    "You have access to the following skills. When a skill is relevant, read its SKILL.md file using the Read tool and follow the instructions within.",
    "",
    ...entries,
    "</available_skills>",
  ].join("\n");
}

export function formatSubAgentsSection(subAgents: SubAgent[]): string {
  let anyAdvisoryMcp = false;
  const entries = subAgents.map((sa) => {
    const parts = [`- **${sa.name}**: ${sa.description}`];
    if (sa.mcpAccess.length > 0) {
      anyAdvisoryMcp = true;
      const servers = sa.mcpAccess.map((a) => a.mcpServer).join(", ");
      parts.push(`  MCP access (advisory): ${servers}`);
    }
    if (sa.modelOverride) {
      parts.push(`  Model: ${sa.modelOverride}`);
    }
    return parts.join("\n");
  });

  const lines = [
    "<sub_agent_delegation>",
    "You can delegate tasks to these specialized sub-agents using the Task tool",
    "(pass the sub-agent's name as the subagent type). They are registered and",
    "run independently, each with its own fresh context.",
    "",
    "Available sub-agents:",
    "",
    ...entries,
    "",
    "Delegation rules:",
    "- Delegate a task to the sub-agent whose specialization matches it.",
    "- Give a clear, self-contained task description — sub-agents do not share your conversation context.",
    "- Sub-agents run independently and return their results when done.",
  ];

  if (anyAdvisoryMcp) {
    lines.push(
      "- \"MCP access (advisory)\" lists the tools a sub-agent is intended to use; " +
        "sub-agents inherit this agent's tool access, so treat it as guidance, not a hard limit.",
    );
  }

  lines.push("</sub_agent_delegation>");
  return lines.join("\n");
}

/**
 * Guidance encouraging the agent to use Cursor's built-in `explore` sub-agent
 * (via the Task tool) for codebase investigation, rather than reading many
 * files inline. Included only when the agent has a workspace to explore.
 *
 * This mirrors how the Cursor IDE aggressively delegates exploration, which the
 * headless agent otherwise rarely does. Kept concise to limit token overhead
 * and paired with a "do not over-delegate" guard to avoid trivial spawns.
 */
export function formatExplorationGuidance(): string {
  return [
    "<codebase_exploration>",
    "For non-trivial investigation of this codebase, prefer delegating to the",
    "built-in `explore` sub-agent via the Task tool instead of reading many",
    "files yourself. Launch one explore task per distinct area you need to",
    "understand — they run in parallel, return focused findings, and keep your",
    "main context clean.",
    "",
    "Use explore for: locating where functionality lives, tracing how a feature",
    "works across files, or surveying unfamiliar areas. Do NOT delegate trivial",
    "single-file reads or small edits you can do directly.",
    "</codebase_exploration>",
  ].join("\n");
}

export function formatWorkspaceContext(dirs: string[]): string {
  if (dirs.length === 1) {
    return `<workspace>\nWorking directory: ${dirs[0]}\n</workspace>`;
  }

  const entries = dirs.map((d, i) => `${i + 1}. ${d}`);
  return [
    "<workspace>",
    "Multi-root workspace with the following directories:",
    ...entries,
    "</workspace>",
  ].join("\n");
}

export function formatInputFiles(
  attachments: readonly AttachmentPromptEntry[],
  vision?: VisionPromptInfo,
): string {
  // A duplicate-renamed file (attachment-naming.ts) discloses its original
  // name so the agent can connect "the two report.pdfs" in the user's
  // message to distinct files on disk.
  const entries = attachments.map((a) =>
    a.renamedFrom !== undefined
      ? `- \`${a.path}\` (renamed from duplicate '${a.renamedFrom}')`
      : `- \`${a.path}\``,
  );
  // The vision lines (shared wording, attachment-vision.ts) tell the model
  // which of these files it can already SEE inline versus which degraded to
  // path-only — without them an agent silently ignores a photo the user
  // believes it can see.
  const disclosure = vision
    ? visionDisclosureLines(vision.inlineFilenames, vision.notViewable)
    : [];
  return [
    "<input_files>",
    "The following files have been provided as inputs. Read them when relevant to the task:",
    ...entries,
    ...disclosure,
    "</input_files>",
  ].join("\n");
}

export function formatReferencedFiles(refs: string[]): string {
  const entries = refs.map((r) => `- \`${r}\``);
  return [
    "<referenced_files>",
    "The user has referenced the following workspace files. Read them when relevant:",
    ...entries,
    "</referenced_files>",
  ].join("\n");
}

/**
 * Returns a system-level directive when the execution is in Plan mode.
 * Returns `undefined` for Agent mode (default) since no prefix is needed.
 *
 * The directive body is the harness-agnostic {@link PLAN_MODE_DIRECTIVE}
 * (shared with the native harness) wrapped in this harness's XML-tag section
 * framing. For Cursor this prompt IS the plan-mode enforcement — the Cursor
 * SDK exposes no mode parameter — see the shared module's doc comment.
 */
export function formatInteractionModePrefix(
  mode: InteractionMode | undefined,
): string | undefined {
  if (mode !== InteractionMode.PLAN) {
    return undefined;
  }

  return ["<interaction_mode>", PLAN_MODE_DIRECTIVE, "</interaction_mode>"].join(
    "\n",
  );
}

/**
 * Returns the implement-plan directive section for a Build-from-plan
 * execution, or `undefined` otherwise. The directive body is the shared
 * {@link buildImplementPlanDirective} (same words as the native harness)
 * wrapped in this harness's XML-tag framing; the variant is picked by whether
 * the approved plan document is among the resolved attachment paths.
 *
 * Like the interaction-mode prefix, this rides BOTH prompt paths — the fresh
 * enhanced prompt and a resumed turn's prefix — because build_from_plan is
 * per-execution and the build turn is usually a follow-up on a resumed agent.
 */
export function formatImplementPlanSection(
  buildFromPlan: boolean | undefined,
  attachments: readonly AttachmentPromptEntry[],
): string | undefined {
  if (!buildFromPlan) {
    return undefined;
  }

  const directive = buildImplementPlanDirective(
    findApprovedPlanPath(attachments.map((a) => a.path)),
  );
  return ["<implement_plan>", directive, "</implement_plan>"].join("\n");
}

export function formatResponseRules(): string {
  return [
    "<response_rules>",
    "- Be concise and direct in your responses",
    "- When making code changes, explain what you changed and why",
    "- If a task is unclear, ask for clarification before proceeding",
    "- Prefer editing existing files over creating new ones",
    "</response_rules>",
  ].join("\n");
}

/**
 * The platform's tool-approval protocol, injected into every Cursor execution.
 *
 * Why this exists ONLY in the Cursor harness (the native LangGraph harness has
 * no such guidance): the two harnesses gate tools differently.
 *
 * - Native gates at the framework level — it calls LangGraph `interrupt()`
 *   BEFORE a tool runs, so the model never sees a denial and cannot react to it
 *   in prose. It also binds only MCP tool descriptions, never a server's
 *   `initialize` `instructions`, so server-authored "ask the user first"
 *   guidance never reaches the model.
 * - Cursor gates via hooks (allow/deny; `ask` is unreliable). The model has
 *   agency over whether it invokes a tool, and Cursor surfaces a connected MCP
 *   server's `instructions` to the model. Servers like open-computer-use ship
 *   "Ask the user before externally-visible actions such as sending…", which
 *   makes the model describe the action in prose and end its turn instead of
 *   invoking the tool. With no tool call, the deny-based gate never fires and
 *   the execution wrongly completes.
 *
 * This directive resolves that conflict in the platform's favor: the agent must
 * invoke tools directly and rely on the platform's automatic approval step,
 * rather than asking for permission in prose. It is the only lever Cursor's
 * hook model leaves available.
 */
const TOOL_APPROVAL_PROTOCOL_INTRO =
  "You run inside a platform that automatically gates sensitive actions for human approval.";

/**
 * The platform's tool-approval rules, authored once and rendered in two places:
 * the system prompt (see {@link formatToolApprovalProtocol}) and an always-applied
 * Cursor project rule (see {@link buildToolApprovalRuleFile}). Keeping a single
 * source means the two surfaces can never drift.
 *
 * The fourth rule is the load-bearing fix for the Cursor harness: because the
 * deny-based gate surfaces an approval pause to the model as a tool *failure*
 * (often Cursor's own generic "blocked by a hook" text — confirmed in Phase 0,
 * and unavoidable since the SDK exposes no non-leaky approval primitive), a
 * well-behaved model otherwise concludes the environment is broken and tells the
 * user to "enable hooks in your Cursor settings", contradicting the approval
 * card. This rule reframes that signal as the gate working as designed.
 */
const TOOL_APPROVAL_PROTOCOL_RULES: readonly string[] = [
  "Carry out every action by calling the appropriate tool directly. Never describe an action you intend to take and then stop, and never ask the user for permission in prose.",
  "When an action needs approval, the platform pauses it, asks the user, and resumes you automatically after they decide. You do not request approval yourself — invoking the tool is how you request it.",
  "Even if a tool or MCP server instructs you to confirm with the user before acting (for example before sending, deleting, or purchasing), do NOT ask in prose. Invoke the tool and let the platform's approval step handle it.",
  "A tool result that says the action was blocked, denied, requires approval, or was \"blocked by a hook\" is the platform's approval gate doing its job — it is NOT an error and NOT a Cursor misconfiguration. Never tell the user to change Cursor settings, enable hooks, or fix their configuration; the gate is intentional and the platform will resume you automatically once the user decides.",
  "If an action is declined, do not retry it or attempt a workaround for it; continue with the rest of the task.",
];

export function formatToolApprovalProtocol(): string {
  return [
    "<tool_approval_protocol>",
    TOOL_APPROVAL_PROTOCOL_INTRO,
    "Follow these rules without exception:",
    ...TOOL_APPROVAL_PROTOCOL_RULES.map((rule) => `- ${rule}`),
    "</tool_approval_protocol>",
  ].join("\n");
}

/**
 * Render the tool-approval protocol as an always-applied Cursor project rule
 * (`.cursor/rules/*.mdc`).
 *
 * A `.cursor/rules` entry with `alwaysApply: true` is injected into the model's
 * context on EVERY turn and takes precedence over a connected MCP server's
 * `instructions`. The system-prompt copy ({@link formatToolApprovalProtocol})
 * only reliably reaches the FIRST user message; the rule file is the durable,
 * higher-precedence surface that keeps the protocol in force across resumed
 * turns and against server-authored "ask the user first" guidance. The runner
 * installs it into the workspace for the duration of a turn and restores the
 * workspace afterward (see workspace-setup.ts).
 */
export function buildToolApprovalRuleFile(): string {
  return [
    "---",
    "description: Stigmer platform tool-approval protocol — how human-in-the-loop approvals work and why a denied/blocked tool is not an error",
    "alwaysApply: true",
    "---",
    "",
    "# Tool approval protocol",
    "",
    TOOL_APPROVAL_PROTOCOL_INTRO,
    "",
    ...TOOL_APPROVAL_PROTOCOL_RULES.map((rule) => `- ${rule}`),
    "",
  ].join("\n");
}
